ddl/syncer/state_syncer.go

```diff
@@ -0,0 +1,198 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package syncer
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/ddl/util"
+	"github.com/pingcap/tidb/metrics"
+	tidbutil "github.com/pingcap/tidb/util"
+	"github.com/pingcap/tidb/util/logutil"
+	"go.etcd.io/etcd/api/v3/mvccpb"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.etcd.io/etcd/client/v3/concurrency"
+	"go.uber.org/zap"
+)
+
+const (
+	// keyOpDefaultTimeout is the default time out for etcd store.
+	keyOpDefaultTimeout = 1 * time.Second
+	statePrompt         = "global-state-syncer"
+	// StateUpgrading represents the cluster global state is upgrading. It is exports for testing.
+	StateUpgrading = "upgrading"
+	// StateNormalRunning represents the cluster global state is normal running. It is exports for testing.
+	StateNormalRunning = ""
+)
+
+// StateSyncer is used to synchronize schema version between the DDL worker leader and followers through etcd.
+type StateSyncer interface {
+	// Init sets the global schema version path to etcd if it isn't exist,
+	// then watch this path, and initializes the self schema version to etcd.
+	Init(ctx context.Context) error
+	// UpdateGlobalState updates the latest version to the global path on etcd until updating is successful or the ctx is done.
+	UpdateGlobalState(ctx context.Context, stateInfo *StateInfo) error
+	// GetGlobalState gets the global state from etcd.
+	GetGlobalState(ctx context.Context) (*StateInfo, error)
+	// IsUpgradingState returns whether the cluster state is upgrading.
+	IsUpgradingState() bool
+	WatchChan() clientv3.WatchChan
+	Rewatch(ctx context.Context)
+}
+
+// StateInfo is the tidb cluster state.
+// It will not be updated when the tidb cluster upgrading.
+type StateInfo struct {
+	State   string `json:"state"`
+	Version string `json:"version"`
```

Comment:
What is the role of version here? For tidb version or something else? If we needn't it in this PR, maybe add it later?

Reply:
Done.